firebase-installations/src/androidTest/java/com/google/firebase/installations/FirebaseInstallationsInstrumentedTest.java

```diff
@@ -217,4 +275,217 @@ public void testGetId_PersistedFidError_BackendOk() throws InterruptedException
+      fail();
+    } catch (ExecutionException expected) {
+      Throwable cause = expected.getCause();
+      assertWithMessage("Exception class doesn't match")
+          .that(cause)
+          .isInstanceOf(FirebaseInstallationsException.class);
+      assertWithMessage("Exception status doesn't match")
+          .that(((FirebaseInstallationsException) cause).getStatus())
+          .isEqualTo(FirebaseInstallationsException.Status.SDK_INTERNAL_ERROR);
+    }
+  }
+
+  @Test
+  public void testGetAuthToken_multipleCalls_fetchedNewTokenOnce() throws Exception {
+    when(mockPersistedFid.readPersistedFidEntryValue())
+        .thenReturn(
+            EXPIRED_AUTH_TOKEN_ENTRY,
+            EXPIRED_AUTH_TOKEN_ENTRY,
+            EXPIRED_AUTH_TOKEN_ENTRY,
+            UPDATED_AUTH_TOKEN_FID_ENTRY);
+    FirebaseInstallations firebaseInstallations =
+        new FirebaseInstallations(
+            mockClock, executor, firebaseApp, backendClientReturnsOk, mockPersistedFid, mockUtils);
+
+    // Call getAuthToken multiple times with DO_NOT_FORCE_REFRESH option
+    Task<InstallationTokenResult> task1 =
+        firebaseInstallations.getAuthToken(FirebaseInstallationsApi.DO_NOT_FORCE_REFRESH);
+    Task<InstallationTokenResult> task2 =
+        firebaseInstallations.getAuthToken(FirebaseInstallationsApi.DO_NOT_FORCE_REFRESH);
+
+    Tasks.await(Tasks.whenAllComplete(task1, task2));
+
+    assertWithMessage("Persisted Auth Token doesn't match")
+        .that(task1.getResult().getToken())
+        .isEqualTo(TEST_AUTH_TOKEN_2);
+    assertWithMessage("Persisted Auth Token doesn't match")
+        .that(task2.getResult().getToken())
+        .isEqualTo(TEST_AUTH_TOKEN_2);
+    verify(backendClientReturnsOk, times(1))
+        .generateAuthToken(anyString(), anyString(), anyString(), anyString());
+  }
+
+  @Test
+  public void testGetAuthToken_multipleCallsForceRefresh_fetchedNewTokenTwice() throws Exception {
+    when(mockPersistedFid.readPersistedFidEntryValue()).thenReturn(REGISTERED_FID_ENTRY);
+    // Use a custom ServiceClient to mock the network calls ensuring task1 is not completed
+    // before task2. Hence, we can test multiple calls to getAUthToken() and verify task2 waits for
+    // task1 to complete.
+    ServiceClient serviceClient = new ServiceClient();
+    FirebaseInstallations firebaseInstallations =
+        new FirebaseInstallations(
+            mockClock, executor, firebaseApp, serviceClient, mockPersistedFid, mockUtils);
+
+    // Call getAuthToken multiple times with FORCE_REFRESH option. Also, sleep for 500ms in between
+    // the calls to ensure tasks are called in order.
+    Task<InstallationTokenResult> task1 =
+        firebaseInstallations.getAuthToken(FirebaseInstallationsApi.FORCE_REFRESH);
+    Thread.sleep(500);
+    Task<InstallationTokenResult> task2 =
+        firebaseInstallations.getAuthToken(FirebaseInstallationsApi.FORCE_REFRESH);
```

Comment:
I wanted the sleep to ensure task executed in order. But I removed it, instead of validating on both expected values

Reply:
Why don't they execute in order at the moment? Is it because the executor has 4 threads? Could we use a single threaded executor to make this deterministic?